man/usnjls.1
.TH USNJLS 1
.SH NAME
usnjls \- List the contents of a NTFS Update Sequence Number journal
.SH SYNOPSIS
.B usnjls [-f
.I fstype
.B ] [-vV]  [-i imgtype] [-o imgoffset] [-b dev_sector_size]
.I image [images] [inode]

.SH DESCRIPTION
.B usnjls
lists the records in a NTFS Update Sequence Number Journal.
If inode is given, then it will look there for a journal.
Otherwise, it will use the default location.
The output lists the USN journal records.

.SH ARGUMENTS
.IP "-f fstype"
Specify the file system type.
Use '\-f list' to list the supported file system types. If not given, autodetection methods are used.
.IP "-i imgtype"
Identify the type of image file, such as raw or split.  Use '\-i list' to list the supported types. If not given, autodetection methods are used.
.IP "-o imgoffset"
The sector offset where the file system starts in the image.
.IP "-b dev_sector_size"
The size, in bytes, of the underlying device sectors.  If not given, the value in the image format is used (if it exists) or 512-bytes is assumed.
.IP -l
Print the output in long format describing the field values and unpacking the data into human readable strings.
.IP -m
Print the output in mactime format.
.IP -V
Display version
.IP -v
verbose output
.IP "image [images]"
One (or more if split) disk or partition images whose format is given with '\-i'.
.IP [inode]
The inode where the Update Sequence Number Journal can be found.

.SH "EXAMPLES"

usnjls \-f ntfs img.dd

.SH AUTHOR
Brian Carrier <carrier at sleuthkit dot org>

Send documentation updates to <doc-updates at sleuthkit dot org>


NEWS.txt
Numbers refer to SourceForge.net tracker IDs:
    http://sourceforge.net/tracker/?group_id=55685

---------------- VERSION 4.3.0 --------------
- PostgreSQL support (Windows only)
- New Release_ NoLibs Visual Studio target
- Support for virtual machine formats via libvmdk and libvhdi (Windows only)
- Schema updates (data sources table, mime type, attributes store type)
- tsk_img_open can take externally created TSK_IMG_INFO
- Various minor bug fixes


---------------- VERSION 4.2.0 --------------
- ExFAT support added
- New database schema
- New Sqlite hash database
- Various bug fixes
- NTFS pays more attention to sequence and loads metadata only 
  if it matches. 
- Added secondary hash database index 



---------------- VERSION 4.1.3 --------------
- fixed bug that could crash UFS/ExtX in inode_lookup.
- More bounds checking in ISO9660 code 
- Image layer bounds checking
- Update version of SQLITE-JDBC
- changed how java loads navite libraries
- Config file for YAFFS2 spare area
- New method in image layer to return names
- Yaffs2 cleanup.
- Escape all strings in SQLite database
- SQlite code uses NTTFS sequence number to match parent IDs


---------------- VERSION 4.1.2 --------------
Core:
- Fixed more visual studio projects to work on 64-bit
- TskAutoDB considers not finding a VS/FS a critical error.

Java: 
- added method to Image to perform sanity check on image sizes.

fiwalk:
- Fixed compile error on Linux etc.



---------------- VERSION 4.1.1 --------------
Core: 
- Added FILE_SHARE_WRITE to all windows open calls.
- removed unused methods in CRC code that caused compile errors.
- Added NTFS FNAME times to time2 struct in TSK_FS_META to make them 
  easier to access -- should have done this a long time ago!
- fls -m and tsk_gettimes output NTFS FNAME times to output for timelines.
- hfind with EnCase hashsets works when DB is specified (and not only index)
- TskAuto now goes into UNALLOC partitions by default too. 
- Added support to automatically find all Cellebrite raw dump files given
  the name of the first image. 
- Added 64-bit windows targets to VisualStudio files.
- Added NTFS sequence to parent address in directory and directory itself.
- Updated SQLite code to use sequence when finding parent object ID.

Java:
- Java bindings JAR files now have native libraries in them. 
- Logical files are added with a transaction 


---------------- VERSION 4.1.0 --------------
Core:
- Added YAFFS2 support (patch from viaForensics).
- Added Ext4 support (patch from kfairbanks)
- changed all include paths to be 'tsk' instead of 'tsk3'
-- IMPORTANT FOR ALL DEVELOPERS!
 
Framework:
- Added Linux and MAC support.
- Added L01 support.
- Added APIs to find files by name, path and extension.
- Removed deprecated TskFile::getAttributes methods.
- moved code around for AutoBuild tool support.

Java Bindings:
- added DerivedFile datamodel support
- added a public method to Content to add ability to close() its tsk handle before the object is gc'd
- added faster skip() and random seek support to ReadContentInputStream
- refactored datamodel by pushing common methods up to AbstractFile
- fixed minor memory leaks 
- improved regression testing framework for java bindings datamodel


---------------- VERSION 4.0.2 --------------
Core: 
New Features:
- Added fiwalk tool from Simson.  Not supported in Visual Studio yet.

Bug Fixes:
- Fixed fcat to work on NTFS files (still doesn't support ADS though). 
- Fixed HFS+ support in tsk_loaddb / SQLite -- root directory was not added.
- NTFS code now looks at all MFT entries when listing directory contents. It used to only look at unallocated entries for orphan files. This fixes an image that had allocated files missing from the directory b-tree.
- NTFS code uses sequence number when searching MFT entries for all files. 
- Libewf detection code change to support v2 API more reliably (ID: 3596212). 
- NTFS $SII code could crash in rare cases if $SDS was multiple of block size. 

Framework:
- Added new API to TskImgDB that returns the base name of an image.
- Numerous performance improvements to framework.
- Removed requirement in framework to specify module extension in pipeline configuration file.
- Added blackboard artifacts to represent both operating system and network service user accounts.

Java Bindings:
- added more APIs to find files by name, path and where clause
- added API to get currently processed dir when image is being added,
- added API to return specific types of children of image, volume system, volume, file system.
- moved more common methods up to Content interface
- deprecated context of blackboard attributes, 
- deprecated SleuthkitCase.runQuery() and SleuthkitCase.closeRunQuery() 
- fixed ReadContentInputStream bugs  (ignoring offset into a buffer, implementing available() )
- methods that are lazy loading are now thread safe
- Hash class is now thread-safe
- use more PreparedStatements to improve performance
- changed source level from java 1.6 to 1.7
- Throw exceptions from C++ side better


---------------- VERSION 4.0.1 --------------
New Features:
- Can open raw Windows devices with write mode sharing.
- More DOS partition types are displayed.
- Added fcat tool that takes in file name and exports content (equivalent to using ifind and icat together).
- Added new API to TskImgDB that returns hash value associated with carved files.
- performance improvements with FAT code (maps and dir_add)
- performance improvements with NTFS code (maps)
- added AONLY flag to block_walk
- Updated blkls and blkcalc to use AONLY flag -- MUCH faster. 

Bug Fixes:
- Fixed mactime issue where it could choose the wrong timezone that did
  not follow daylight savings times. 
- Fixed file size of alternate data streams in framework.
- Incorporated memory leak fixes and raw device fixes from ADF Solutions.



---------------- VERSION 4.0.0 --------------
New Features:
- Added multithreaded support
- Added C++ wrapper classes

tools/fstools/usnjls.cpp
/*
** usnjls
** The Sleuth Kit
**
** Given an NTFS image and UsnJrnl file inode, parses its content showing
** the list of recent changes wihtin the File System.
**
** Matteo Cafasso [noxdafox <at> gmail [dot] com]
**
** This software is distributed under the Common Public License 1.0
**
*/


#include <locale.h>
#include "tsk/tsk_tools_i.h"


static TSK_TCHAR *progname;


/* usage - explain and terminate */
static void
usage()
{
    TFPRINTF(stderr,
             _TSK_T
             ("usage: %s [-f fstype] [-i imgtype] [-b dev_sector_size]"
              " [-o imgoffset] [-lmvV] image inode\n"),
             progname);
    tsk_fprintf(stderr,
                "\t-i imgtype: The format of the image file "
                "(use '-i list' for supported types)\n");
    tsk_fprintf(stderr,
                "\t-b dev_sector_size: The size (in bytes)"
                " of the device sectors\n");
    tsk_fprintf(stderr,
                "\t-f fstype: File system type "
                "(use '-f list' for supported types)\n");
    tsk_fprintf(stderr,
                "\t-o imgoffset: The offset of the file system"
                " in the image (in sectors)\n");
    tsk_fprintf(stderr, "\t-l: Long output format with detailed information\n");
    tsk_fprintf(stderr, "\t-m: Time machine output format\n");
    tsk_fprintf(stderr, "\t-v: verbose output to stderr\n");
    tsk_fprintf(stderr, "\t-V: print version\n");

    exit(1);
}


int
main(int argc, char **argv1)
{
    TSK_IMG_INFO *img = NULL;
    TSK_IMG_TYPE_ENUM imgtype = TSK_IMG_TYPE_DETECT;

    TSK_FS_INFO *fs = NULL;
    TSK_OFF_T imgaddr = 0;
    TSK_FS_FILE *jrnl_file = NULL;
    TSK_FS_TYPE_ENUM fstype = TSK_FS_TYPE_DETECT;

    int ch;
    TSK_INUM_T inum;
    TSK_TCHAR *cp = NULL;
    unsigned int ssize = 0;
    TSK_TCHAR **argv = NULL;
    TSK_FS_USNJLS_FLAG_ENUM flag = TSK_FS_USNJLS_NONE;

#ifdef TSK_WIN32
    // On Windows, get the wide arguments (mingw doesn't support wmain)
    argv = CommandLineToArgvW(GetCommandLineW(), &argc);
    if (argv == NULL) {
        fprintf(stderr, "Error getting wide arguments\n");
        exit(1);
    }
#else
    argv = (TSK_TCHAR **) argv1;
#endif

    progname = argv[0];
    setlocale(LC_ALL, "");

    while ((ch = GETOPT(argc, argv, _TSK_T("b:f:i:o:lmvV"))) > 0) {
        switch (ch) {
        case _TSK_T('?'): {
            default:
                TFPRINTF(stderr, _TSK_T("Invalid argument: %s\n"),
                         argv[OPTIND]);
                usage();
        }
        case _TSK_T('b'):
            ssize = (unsigned int) TSTRTOUL(OPTARG, &cp, 0);
            if (*cp || *cp == *OPTARG || ssize < 1) {
                TFPRINTF(stderr,
                         _TSK_T("invalid argument: sector size "
                                "must be positive: %s\n"),
                         OPTARG);
                usage();
            }
            break;
        case _TSK_T('f'):
            if (TSTRCMP(OPTARG, _TSK_T("list")) == 0) {
                tsk_fs_type_print(stderr);
                exit(1);
            }
            fstype = tsk_fs_type_toid(OPTARG);
            if (fstype == TSK_FS_TYPE_UNSUPP) {
                TFPRINTF(stderr,
                         _TSK_T("Unsupported file system type: %s\n"), OPTARG);
                usage();
            }
            break;
        case _TSK_T('i'):
            if (TSTRCMP(OPTARG, _TSK_T("list")) == 0) {
                tsk_img_type_print(stderr);
                exit(1);
            }
            imgtype = tsk_img_type_toid(OPTARG);
            if (imgtype == TSK_IMG_TYPE_UNSUPP) {
                TFPRINTF(stderr, _TSK_T("Unsupported image type: %s\n"),
                         OPTARG);
                usage();
            }
            break;
        case _TSK_T('o'):
            if ((imgaddr = tsk_parse_offset(OPTARG)) == -1) {
                tsk_error_print(stderr);
                exit(1);
            }
            break;
        case _TSK_T('l'):
            flag = TSK_FS_USNJLS_LONG;
            break;
        case _TSK_T('m'):
            flag = TSK_FS_USNJLS_MAC;
            break;
        case _TSK_T('v'):
            tsk_verbose++;
            break;
        case _TSK_T('V'):
            tsk_version_print(stdout);
            exit(0);
        }
    }

    /* We need at least one more argument */
    if (OPTIND >= argc) {
        tsk_fprintf(stderr, "Missing image name and/or address\n");
        usage();
    }

    /* open image - there is an optional inode address at the end of args.
     *
     * Check the final argument and see if it is a number
     */
    if (tsk_fs_parse_inum(argv[argc - 1], &inum, NULL, NULL, NULL, NULL)) {
        img = tsk_img_open(argc - OPTIND, &argv[OPTIND], imgtype, ssize);
        if (img == NULL) {
            tsk_error_print(stderr);
            exit(1);
        }

        if ((imgaddr * img->sector_size) >= img->size) {
            tsk_fprintf(stderr,
                        "Sector offset is larger than disk image (maximum: %"
                        PRIu64 ")\n", img->size / img->sector_size);
            exit(1);
        }

        fs = tsk_fs_open_img(img, imgaddr * img->sector_size, fstype);
        if (fs == NULL) {
            tsk_error_print(stderr);

            if (tsk_error_get_errno() == TSK_ERR_FS_UNSUPTYPE)
                tsk_fs_type_print(stderr);

            img->close(img);
            exit(1);
        }

        jrnl_file = tsk_fs_file_open(fs, NULL, "$Extend/$UsnJrnl");
        if (jrnl_file == NULL) {
            tsk_fprintf(stderr,
                        "Unable to open USN Journal file $Extend/$UsnJrnl\n");

            exit(1);
        }

        inum = jrnl_file->name->meta_addr;
        free(jrnl_file);
    } else {
        img = tsk_img_open(argc - OPTIND - 1, &argv[OPTIND], imgtype, ssize);
        if (img == NULL) {
            tsk_error_print(stderr);
            exit(1);
        }

        if ((imgaddr * img->sector_size) >= img->size) {
            tsk_fprintf(stderr,
                        "Sector offset is larger than disk image (maximum: %"
                        PRIu64 ")\n", img->size / img->sector_size);
            exit(1);
        }

        fs = tsk_fs_open_img(img, imgaddr * img->sector_size, fstype);
        if (fs == NULL) {
            tsk_error_print(stderr);

            if (tsk_error_get_errno() == TSK_ERR_FS_UNSUPTYPE)
                tsk_fs_type_print(stderr);

            img->close(img);
            exit(1);
        }
    }

    if (inum > fs->last_inum) {
        tsk_fprintf(stderr,
                    "Inode value is too large for image (%" PRIuINUM ")\n",
                    fs->last_inum);
        fs->close(fs);
        img->close(img);
        exit(1);
    }

    if (inum < fs->first_inum) {
        tsk_fprintf(stderr,
                    "Inode value is too small for image (%" PRIuINUM ")\n",
                    fs->first_inum);
        fs->close(fs);
        img->close(img);
        exit(1);
    }

    if (tsk_fs_usnjls(fs, inum, flag)) {
        tsk_error_print(stderr);
        fs->close(fs);
        img->close(img);
        exit(1);
    }

    fs->close(fs);
    img->close(img);
    exit(0);
}
